src/lib.rs

```diff
@@ -95,161 +95,111 @@ impl Slice {
-            end.map_or(Bound::Included(0), Bound::Excluded)
-        };
-        RangeIterator::new(start, end, step)
     }
 }
 
 impl Index {
-    /// absolute index. negative indices are added to len.
-    fn abs(&self, len: usize, dir: &Direction) -> Option<usize> {
+    fn to_signed(&self) -> Option<i128> {
         match self {
-            &Head(n) => ensure_within(n, 0..len),
-            // TODO: simplify/cleanup
-            &Tail(n) => match dir {
-                Forwards => Some(len - len.min(n)),
-                Backwards => ensure_within(n, 1..len + 1).map(|n| len - n),
-            },
+            &Head(n) => Some(n as i128),
+            &Tail(n) => Some(-(n as i128)),
             Default => None,
         }
     }
 }
 
-/// Return Some(n) if r.start <= n < r.end, otherwise return None.
-fn ensure_within(n: usize, r: Range<usize>) -> Option<usize> {
-    if r.contains(&n) {
-        Some(n)
-    } else {
-        None
-    }
+/// A SSlice is a slice implemented using signed arithmetics. It uses a signed type that
+/// must be able to represent usize. This impl won't work if size_of<usize> >= 16.
+/// 64 bits of address space should be enough for everybody.
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct SSlice {
+    start: Option<i128>,
+    end: Option<i128>,
+    step: Option<i128>,
 }
 
-/// An iterator that counts from an initial number up to a (included or excluded) final limit.
-/// The direction and stride of the iteration can be controlled by the step parameter.
-/// A zero step produces an empty iteration.
-#[derive(Debug)]
-struct RangeIterator {
-    end: Bound<usize>,
-    step: isize,
-    pos: usize,
-    done: bool,
-    dir: Direction,
-}
+impl SSlice {
+    /// Returns an iterator that yields the indices that match the slice expression.
+    fn indices(&self, len: usize) -> impl Iterator<Item = usize> {
+        let ilen = len as i128;
 
-impl RangeIterator {
-    fn new(start: usize, end: Bound<usize>, step: isize) -> Self {
-        RangeIterator {
-            pos: start,
-            end,
-            step,
-            done: false,
-            dir: Direction::new(step),
+        if len == 0 {
+            return Iter {
+                i: 0,
+                end: 0,
+                step: 0,
+            };
         }
-    }
-}
 
-// Iteration direction, knows how to compare the limit index.
-#[derive(Debug)]
-enum Direction {
-    Forwards,
-    Backwards,
-}
+        let step = self.step.unwrap_or(1);
 
-impl Direction {
-    fn new(step: isize) -> Self {
-        if step >= 0 {
-            Direction::Forwards
+        let (def_start, def_end) = if step >= 0 {
+            (0, ilen)
         } else {
-            Direction::Backwards
+            (ilen - 1, -ilen - 1)
+        };
+
+        let shift = if step >= 0 { 0 } else { 1 };
+        let min = 0 - shift;
+        let max = ilen - shift;
+
+        let abs = |n: i128| if n >= 0 { n } else { ilen + n };
```

Comment:
If find the name `abs` ("absolute value") slightly confusing since the result can be negative.

Reply:
PTAL

Reply:
Better name, thanks.